Allow access to directories not ending with slash

ayg opened the pull request:
Commit 8016c0f6 returns a 404 in this case.  I don't know why, maybe it
complicates subsequent processing?  This patch just pretends that there
was a slash at the end.  If that's not nice enough, a cleaner solution
is to issue a 301 to the URL with a slash, but I don't know how to do
that offhand.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/6775)
<!-- Reviewable:end -->

ayg commented:
I run into this on a regular basis, because I often type path names manually and don't normally add the slash.  Normal webservers do support this behavior AFAICT.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/261862072)

# Firefox (nightly) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/261862073)

# Sauce (safari) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/261862082)

# Chrome (unstable) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/261862083)

# Sauce (MicrosoftEdge) #

jgraham commented:
This was pretty deliberate, to avoid implicit magic in the server as much as possible, in case it confuses tests. I don't specifically know how this would cause tests to go wrong, but I would at least like more input before merging such a change.

Ms2ger commented:
The issue I remember is that links on the overview page for the directory would not work, because `file` resolved relative to `/foo/bar` rather than `/foo/bar/` yields `/foo/file` rather than `foo/bar/file`.

jgraham commented:
Good memory there.

ayg commented:
So any pointers on how to return a 301?

ayg committed:
Allow access to directories not ending with a slash

This returns a 301 to the version with the slash.  IIRC this matches the
behavior of normal web servers.  Tested locally, seems to work.

ayg commented:
Okay, new version submitted that returns a 301.

ayg committed:
fixup! Allow access to directories not ending with a slash

ayg commented:
Attempted to fix test failures, although I couldn't verify locally because I couldn't get the tests to run.

ayg committed:
fixup! Allow access to directories not ending with a slash

ayg committed:
fixup! Allow access to directories not ending with a slash

ayg commented:
Finally fixed all the test failures.

ayg commented:
Causes no extra test failures in Gecko: https://treeherder.mozilla.org/#/jobs?repo=try&revision=59dce63ed354502893efce3aefd3e1df5c004a3b

w3c-bots commented:
# Build PASSED

Started: 2017-08-14 14:44:35
Finished: 2017-08-14 15:02:46





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/17362)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/264387516)
